File: legacy/src/main/python/preponderous/viron/services/entityService.py
import logging
from typing import List, Optional
import requests

from src.main.python.preponderous.viron.models.entity import Entity

logger = logging.getLogger(__name__)

class EntityService:
    def __init__(self, viron_host: str, viron_port: int):
        self.viron_host = viron_host
        self.viron_port = viron_port

    def get_base_url(self) -> str:
        return f"{self.viron_host}:{self.viron_port}/api/v1/entities"

    def get_all_entities(self) -> List[Entity]:
        response = requests.get(self.get_base_url())
        response.raise_for_status()
        data = response.json()
        return [Entity(**entity) for entity in data] if data else []

    def get_entity_by_id(self, entity_id: int) -> Optional[Entity]:
        response = requests.get(f"{self.get_base_url()}/{entity_id}")
        response.raise_for_status()
        data = response.json()
        return Entity(**data) if data else None

    def get_entities_in_environment(self, environment_id: int) -> List[Entity]:
        response = requests.get(f"{self.get_base_url()}/environment/{environment_id}")
        response.raise_for_status()
        data = response.json()
        return [Entity(**entity) for entity in data] if data else []

    def get_entities_in_grid(self, grid_id: int) -> List[Entity]:
        response = requests.get(f"{self.get_base_url()}/grid/{grid_id}")
        response.raise_for_status()
        data = response.json()
        return [Entity(**entity) for entity in data] if data else []

    def get_entities_in_location(self, location_id: int) -> List[Entity]:
        response = requests.get(f"{self.get_base_url()}/location/{location_id}")
        response.raise_for_status()
        data = response.json()
        return [Entity(**entity) for entity in data] if data else []

    def get_entities_not_in_any_location(self) -> List[Entity]:
        response = requests.get(f"{self.get_base_url()}/unassigned")
        response.raise_for_status()
        data = response.json()
        return [Entity(**entity) for entity in data] if data else []

    def create_entity(self, name: str) -> Entity:
        response = requests.post(f"{self.get_base_url()}/{name}")
        response.raise_for_status()
        data = response.json()
        if not data:
            raise Exception("Created entity response was null")
        return Entity(**data)

    def delete_entity(self, entity_id: int) -> bool:
        try:
            response = requests.delete(f"{self.get_base_url()}/{entity_id}")
            response.raise_for_status()
            return True
        except Exception as e:
            logger.error(f"Failed to delete entity {entity_id}: {str(e)}")
            raise Exception("Error deleting entity", e)

    def update_entity_name(self, entity_id: int, name: str) -> bool:
        try:
            response = requests.patch(f"{self.get_base_url()}/{entity_id}/name/{name}")
            response.raise_for_status()
            return True
        except Exception as e:
            logger.error(f"Failed to update name for entity {entity_id}: {str(e)}")
            raise Exception("Error updating entity name", e)

File: legacy/src/main/java/preponderous/viron/services/EntityService.java
package preponderous.viron.services;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import preponderous.viron.config.ServiceConfig;
import preponderous.viron.exceptions.EntityServiceException;
import preponderous.viron.models.Entity;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

@Service
@Slf4j
@RequiredArgsConstructor
public class EntityService {
    private final RestTemplate restTemplate;
    private final ServiceConfig serviceConfig;

    private String getBaseUrl() {
        return String.format("%s:%d/api/v1/entities",
                serviceConfig.getVironHost(),
                serviceConfig.getVironPort());
    }

    public List<Entity> getAllEntities() {
        try {
            ResponseEntity<Entity[]> response = restTemplate.getForEntity(
                    getBaseUrl(),
                    Entity[].class
            );

            return response.getBody() != null
                    ? Arrays.asList(response.getBody())
                    : Collections.emptyList();

        } catch (Exception e) {
            log.error("Failed to fetch all entities: {}", e.getMessage());
            throw new EntityServiceException("Error retrieving entities", e);
        }
    }

    public Optional<Entity> getEntityById(int id) {
        try {
            ResponseEntity<Entity> response = restTemplate.getForEntity(
                    getBaseUrl() + "/{id}",
                    Entity.class,
                    id
            );

            return Optional.ofNullable(response.getBody());

        } catch (Exception e) {
            log.error("Failed to fetch entity with id {}: {}", id, e.getMessage());
            throw new EntityServiceException("Error retrieving entity", e);
        }
    }

    public List<Entity> getEntitiesInEnvironment(int environmentId) {
        try {
            ResponseEntity<Entity[]> response = restTemplate.getForEntity(
                    getBaseUrl() + "/environment/{environmentId}",
                    Entity[].class,
                    environmentId
            );

            return response.getBody() != null
                    ? Arrays.asList(response.getBody())
                    : Collections.emptyList();

        } catch (Exception e) {
            log.error("Failed to fetch entities in environment {}: {}", environmentId, e.getMessage());
            throw new EntityServiceException("Error retrieving entities in environment", e);
        }
    }

    public List<Entity> getEntitiesInGrid(int gridId) {
        try {
            ResponseEntity<Entity[]> response = restTemplate.getForEntity(
                    getBaseUrl() + "/grid/{gridId}",
                    Entity[].class,
                    gridId
            );

            return response.getBody() != null
                    ? Arrays.asList(response.getBody())
                    : Collections.emptyList();

        } catch (Exception e) {
            log.error("Failed to fetch entities in grid {}: {}", gridId, e.getMessage());
            throw new EntityServiceException("Error retrieving entities in grid", e);
        }
    }

    public List<Entity> getEntitiesInLocation(int locationId) {
        try {
            ResponseEntity<Entity[]> response = restTemplate.getForEntity(
                    getBaseUrl() + "/location/{locationId}",
                    Entity[].class,
                    locationId
            );

            return response.getBody() != null
                    ? Arrays.asList(response.getBody())
                    : Collections.emptyList();

        } catch (Exception e) {
            log.error("Failed to fetch entities in location {}: {}", locationId, e.getMessage());
            throw new EntityServiceException("Error retrieving entities in location", e);
        }
    }

    public List<Entity> getEntitiesNotInAnyLocation() {
        try {
            ResponseEntity<Entity[]> response = restTemplate.getForEntity(
                    getBaseUrl() + "/unassigned",
                    Entity[].class
            );

            return response.getBody() != null
                    ? Arrays.asList(response.getBody())
                    : Collections.emptyList();

        } catch (Exception e) {
            log.error("Failed to fetch unassigned entities: {}", e.getMessage());
            throw new EntityServiceException("Error retrieving unassigned entities", e);
        }
    }

    public Entity createEntity(String name) {
        try {
            ResponseEntity<Entity> response = restTemplate.postForEntity(
                    getBaseUrl() + "/{name}",
                    null,
                    Entity.class,
                    name
            );

            if (response.getBody() == null) {
                throw new EntityServiceException("Created entity response was null");
            }

            return response.getBody();

        } catch (Exception e) {
            log.error("Failed to create entity with name {}: {}", name, e.getMessage());
            throw new EntityServiceException("Error creating entity", e);
        }
    }

    public boolean deleteEntity(int id) {
        try {
            restTemplate.delete(getBaseUrl() + "/{id}", id);
            return true;
        } catch (Exception e) {
            log.error("Failed to delete entity {}: {}", id, e.getMessage());
            throw new EntityServiceException("Error deleting entity", e);
        }
    }

    public boolean updateEntityName(int id, String name) {
        try {
            restTemplate.patchForObject(
                    getBaseUrl() + "/{id}/name/{name}",
                    null,
                    Void.class,
                    id,
                    name
            );
            return true;
        } catch (Exception e) {
            log.error("Failed to update name for entity {}: {}", id, e.getMessage());
            throw new EntityServiceException("Error updating entity name", e);
        }
    }
}